cocos/network/CCWebSocket-android.cpp

```diff
@@ -0,0 +1,403 @@
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+
+
+#include "network/CCWebSocket-android.h"
+#include "network/WebSocket.h"
+#include "platform/CCFileUtils.h"
+#include "platform/android/jni/JniHelper.h"
+
+
+#define JCLS_WEBSOCKET "org/cocos2dx/lib/Cocos2dxWebSocket"
+#define J_BINARY_CLS_WEBSOCKET "org.cocos2dx.lib.Cocos2dxWebSocket"
+#define JARG_STR        "Ljava/lang/String;"
+#define JARG_DOWNLOADER "L" JCLS_WEBSOCKET ";"
+
+#include <string>
+#include <mutex>
+#include <unordered_map>
+#include <condition_variable>
+#include <chrono>
+
+using namespace std;
+
+
+namespace {
+
+    class ConditionVariable {
+    public:
+
+        void wait_for(const std::chrono::duration<float>& timeout) {
+            std::unique_lock<std::mutex> lock(_mutex);
+            _cond.wait_for(lock, timeout);
+        }
+
+        void notify_all() {
+            _cond.notify_all();
+        }
+
+    private:
+        std::mutex _mutex;
+        std::condition_variable _cond;
+    };
+
+    class ConnectionCV {
+    public:
+        ConnectionCV() = default;
+
+        virtual ~ConnectionCV() {
+            std::lock_guard<std::mutex> guard(_mtx);
+            _conditions.clear();
+        }
+
+        void watchConnection(int64_t id)
+        {
+            std::lock_guard<std::mutex> guard(_mtx);
+            CCASSERT(_conditions.find(id) == _conditions.end(), "id should not be registered before");
+            auto p = std::make_shared<ConditionVariable>();
+            _conditions.emplace(id, p);
+        }
+
+        void unwatchConnection(int64_t id)
+        {
+            std::lock_guard<std::mutex> guard(_mtx);
+            _conditions.erase(id);
+        }
+
+        void notifyClosed(int64_t id)
+        {
+            std::shared_ptr<ConditionVariable> p = nullptr;
+            {
+                std::lock_guard<std::mutex> guard(_mtx);
+                auto itr = _conditions.find(id);
+                if(itr == _conditions.end())
+                {
+                    return;
+                }
+                p = itr->second;
+            }
+            p->notify_all();
+        }
+
+        void waitForClosing(int64_t id, std::chrono::duration<float> timeout)
+        {
+            std::shared_ptr<ConditionVariable> p = nullptr;
+            {
+                std::lock_guard<std::mutex> guard(_mtx);
+                auto itr = _conditions.find(id);
+                if (itr == _conditions.end()) {
+                    return;
+                }
+                p = itr->second;
+            }
+            p->wait_for(timeout);
+        }
+    private:
+        std::mutex _mtx;
+        std::unordered_map<int64_t, std::shared_ptr<ConditionVariable>> _conditions;
```

Comment:
`shared_ptr` is used to prevent a local reference from being deleted. 

In `waitForClosing`
```cpp
             {
               std::lock_guard<std::mutex> guard(_mtx);
                auto itr = _conditions.find(id);
                if (itr == _conditions.end()) {
                    return;
                }
                p = itr->second;
            }
            // prevent p from being delete here
            p->wait_for(timeout);
```